internal/pkg/flink/internal/controller/table_controller_test.go

```diff
@@ -370,3 +386,61 @@ func (s *TableControllerTestSuite) TestDefaultInRowView() {
 	assert.Equal(s.T(), input, result)
 	assert.True(s.T(), tableController.isRowViewOpen)
 }
+
+func (s *TableControllerTestSuite) TestResultFetchStopsAfterError() {
+	s.runWithRealTView(func(tview *tview.Application) {
+		// Given
+		table := components.CreateTable()
+		mockStatement := types.ProcessedStatement{PageToken: "NOT_EMPTY"}
+		tableController := TableController{
+			table:         table,
+			appController: s.mockAppController,
+			store:         s.mockStore,
+		}
+		tableController.SetRunInteractiveInputCallback(s.mockInputController.RunInteractiveInput)
+		s.mockAppController.EXPECT().TView().Return(tview).Times(4)
+		s.mockStore.EXPECT().FetchStatementResults(mockStatement).Return(&mockStatement, &types.StatementError{Message: "error"})
+
+		// When
+		tableController.Init(mockStatement)
+		assert.True(s.T(), tableController.isAutoRefreshRunning())
+		assert.Equal(s.T(), running, int(atomic.LoadInt32(&tableController.fetchState)))
+		// wait for auto refresh to complete
+		for tableController.isAutoRefreshRunning() {
+			time.Sleep(1 * time.Second)
+		}
+
+		// Then
+		assert.False(s.T(), tableController.isAutoRefreshRunning())
+		assert.Equal(s.T(), failed, int(atomic.LoadInt32(&tableController.fetchState)))
+	})
+}
+
+func (s *TableControllerTestSuite) TestResultFetchStopsAfterNoMorePageToken() {
+	s.runWithRealTView(func(tview *tview.Application) {
+		// Given
+		table := components.CreateTable()
+		mockStatement := types.ProcessedStatement{PageToken: "NOT_EMPTY"}
+		tableController := TableController{
+			table:         table,
+			appController: s.mockAppController,
+			store:         s.mockStore,
+		}
+		tableController.SetRunInteractiveInputCallback(s.mockInputController.RunInteractiveInput)
+		s.mockAppController.EXPECT().TView().Return(tview).Times(4)
```

Comment:
where does the 4 comes from? is it a default value? can we import and use it here?

Reply:
I counted this myself lol. We expect to have two calls to this `t.appController.TView().QueueUpdateDraw(t.renderTable)`. The first is in the default case of the results fetch loop and then the second one happens when the context is cancelled an we enter the `ctx.Done()` case. Each of those calls counts as 2 calls to `TView()` because t.renderTable also calls `TView()` once.

I could declare it as a variable but you would still need to count this again yourself if the amount of calls changes in the future.